### PR TITLE
bpo-23892: Introduce sys.implementation.opt_levels

### PR DESCRIPTION
**Pull request by csabella**
https://bugs.python.org/issue23892

### CODE REVIEW

File: Python/sysmodule.c [hunk 1]
```diff
@@ -2232,6 +2232,11 @@ make_impl_info(PyObject *version_info)
     if (res < 0)
         goto error;
 
+    res = PyDict_SetItemString(impl_info, "opt_levels",
+                               Py_BuildValue("(sii)", "", 1, 2));
```
Comment (ZackerySpytz) [inline Python/sysmodule.c]: The ``Py_BuildValue()`` call needs to be handled like the ``PyLong_FromLong()`` call just above. The current code causes a reference leak (or a crash if the ``Py_BuildValue()`` call fails).

Comment (csabella) [reply Python/sysmodule.c]: Thank you!  I've pushed the change.

File: Doc/library/sys.rst [hunk 2]
```diff
@@ -814,6 +814,11 @@ always available.
    ``cache_tag`` is set to ``None``, it indicates that module caching should
    be disabled.
 
+   *opt_levels* is a tuple containing the available compiler-supported
+   optimization levels.  See :pep:`488` for more information.
```
Comment (taleinat) [inline Doc/library/sys.rst:818]: I think that mentioning the possible values, and the relation to the "optimization" parameter of `importlib.util.cache_from_source()`, is perhaps warranted here (in addition to linking to PEP-488).

**Comment by taleinat** [comment]
Looks solid overall.

File: Doc/library/sys.rst [hunk 3]
```diff
@@ -814,6 +814,11 @@ always available.
...
+
+   .. versionadded:: 3.8
```
Comment (ericsnowcurrently) [inline Doc/library/sys.rst:820]: 3.9 now :)

File: Lib/test/test_sys.py [hunk 4]
```diff
@@ -736,6 +736,7 @@ def test_implementation(self):
         self.assertTrue(hasattr(sys.implementation, 'version'))
         self.assertTrue(hasattr(sys.implementation, 'hexversion'))
         self.assertTrue(hasattr(sys.implementation, 'cache_tag'))
+        self.assertTrue(hasattr(sys.implementation, 'opt_levels'))
```
Comment (ericsnowcurrently) [inline Lib/test/test_sys.py:740]: Hmm, we should probably also make sure that no other required attrs exist.  (We could parameterize this part at the same time.)

```python
required = [attr for attr in dir(sys.implementation) if not attr.startswith('_')]
expected = ['name', 'version', 'hexversion', 'cache_tag', 'opt_levels']
self.assertCountEqual(required, expected)
# Make sure (best effort) that sys.implementation.__dir__() is correct.
for attr in expected:
    self.assertTrue(hasattr(sys.implementation, attr))
```

We should also make sure that `sys.implementation.opt_levels` is a tuple of only non-negative ints and alphanumeric strings.

```python
self.assertIs(type(sys.implementation.opt_levels), tuple, sys.implementation.opt_levels)
for opt in sys.implementation.opt_levels:
    if type(opt) is int:
        self.assertTrue(opt.isalnum(), opt)
    else:
        self.assertIs(type(opt), int, opt)
        self.assertGreaterEqual(opt, 0)
```

File: Lib/zipfile.py [hunk 5]
```diff
@@ -2034,63 +2034,44 @@ def _compile(file, optimize=-1):
                 return False
             return True
 
+        def _check_cache(pycache, file):
```
Comment (ericsnowcurrently) [inline Lib/zipfile.py:2037]: nice :)

File: Lib/zipfile.py [hunk 6]
```diff
@@ -2034,63 +2034,44 @@ def _compile(file, optimize=-1):
...
+                        # Use the __pycache__/*.pyc file, but write it to the legacy pyc
+                        # file name in the archive.
+                        fname = pycache
+                        arcname = file_pyc
+                        break
                 else:
-                    fname = arcname = file_py
+                    # Compile py into PEP 3147 pyc file.
+                    if _compile(file_py):
+                        fname = pycache_opt[sys.flags.optimize]
```
Comment (ericsnowcurrently) [inline Lib/zipfile.py:2062]: Shouldn't this catch the IndexError and fall back `fname = arcname = file_py`?

FWIW, this code implies that `sys.flags.optimize` is guaranteed to be suitable as a tuple index (but such a guarantee does not exist).

Also, PEP 488 opens that door for dealing with sets of optimizations (either as a container of strings/ints or as a hash of them).  This code would fall apart then.

We can discuss this more on the BPO issue.

File: Lib/zipfile.py [hunk 7]
```diff
@@ -2034,63 +2034,44 @@ def _compile(file, optimize=-1):
...
-                    else:
-                        fname = pycache_opt2
-                    arcname = file_pyc
+                for pycache in pycache_opt:
+                    if _check_cache(pycache, file_py):
+                        # Use the __pycache__/*.pyc file, but write it to the legacy pyc
+                        # file name in the archive.
+                        fname = pycache
+                        arcname = file_pyc
+                        break
```
Comment (ericsnowcurrently) [inline Lib/zipfile.py:2058]: This "break" implies that the order of `sys.implementation.opt_levels` is extremely important.  In a world where we deal only in sets of optimizations represented as increasing interger "levels" thats fine, but PEP 488 suggests that's not necessarily where we're headed.  We should discuss this more on the BPO issue. :)

If we end up sticking with opt levels as currently used then the importance of order should be made clear in the docs.

**Review by ericsnowcurrently — COMMENTED** [review]
As I noted on the BPO issue, I think there are problems with the naive approach I suggested.  We should discuss more there.

Also, I'd expect changes to `Lib/importlib/_bootstrap_external.py`...

Also, didn't zipimport get rewritten recently?

File: Doc/library/sys.rst (hunk 2 above)
Comment (ericsnowcurrently) [reply Doc/library/sys.rst:818]: +1

The relationship between `sys.flags.optimize` and that "optimization" parameter (as well as the "optimize" parameter of `compile()`) reflects a gap in the solution presented by PEP 488.  That needs to be resolved.

File: Lib/zipfile.py (hunk 5 above)
Comment (csabella) [reply Lib/zipfile.py:2037]: Thanks!  I think even if this ends up not being in sys.implementation, these changes in zipfile still might be worth adding to simplify the code.

**Pull request closed without merging**